source/extensions/geoip_providers/maxmind/geoip_provider.h

```diff
@@ -103,14 +103,12 @@ class GeoipProvider : public Envoy::Geolocation::Driver,
   // Allow the unit test to have access to private members.
   friend class GeoipProviderPeer;
   GeoipProviderConfigSharedPtr config_;
-  // These three mutexes are never held at the same time. We signify this by requiring
-  // that all three are 'acquired before' the others, since there is no exclusion annotation.
-  mutable absl::Mutex city_db_mutex_ ABSL_ACQUIRED_BEFORE(isp_db_mutex_, anon_db_mutex_){};
-  MaxmindDbSharedPtr city_db_;
-  mutable absl::Mutex isp_db_mutex_ ABSL_ACQUIRED_BEFORE(city_db_mutex_, anon_db_mutex_){};
-  MaxmindDbSharedPtr isp_db_;
-  mutable absl::Mutex anon_db_mutex_ ABSL_ACQUIRED_BEFORE(city_db_mutex_, isp_db_mutex_){};
-  MaxmindDbSharedPtr anon_db_;
+  mutable absl::Mutex city_db_mutex_;
+  MaxmindDbSharedPtr city_db_ ABSL_GUARDED_BY(city_db_mutex_);
+  mutable absl::Mutex isp_db_mutex_;
+  MaxmindDbSharedPtr isp_db_ ABSL_GUARDED_BY(isp_db_mutex_);
+  mutable absl::Mutex anon_db_mutex_;
```

Comment:
Previous comment was "why not use one mutex if they're never acquired at the same time"; now we've lost the ACQUIRED_BEFORE annotation that meant there could be no deadlocks, but still have 3 mutexes so now it's in danger of accidental deadlock.

Again, why not just have one mutex guarding all three things?

Reply:
Apologies for slow response on this, did not have much capacity to work on it in the past week. I think we can switch to using one mux, since reload logic runs on a single custom thread (since geoip provider instance is unique per config hash), so only one file reload event will be processed at a time. And all writes to those mux guarded fields happen on that custom thread, except of the case when provider is constructed and dbs are loaded for the first time. But the constructor code is structured in a way so that all db initial loads happen before custom thread is allocated. Since in our case the dbs are reloaded infrequently (this is the expectation, Maxmind publishes new datasets once week) but are heavily being read, we could single mutex and lock it in [reader/writer mode](https://abseil.io/docs/cpp/guides/synchronization#reader-writer-locks).

Reply:
One downside to this approach:
```
Note that ReaderLock() may block even if the lock
  // is held in shared mode; this occurs when another thread is blocked on a
  // call to WriterLock().
  ```
  From: https://github.com/abseil/abseil-cpp/blob/master/absl/synchronization/mutex.h#L220C32-L222C27
  So it means  if one db is being reloaded all reads to all db types will be affected.

Reply:
nit: just wanted to add that for certain maxmind database such as the anonymous IP database it could be updated daily so slightly more frequently (https://support.maxmind.com/hc/en-us/articles/4408216129947-Download-and-Update-Databases)

Reply:
> So it means if one db is being reloaded all reads to all db types will be affected.

This should be fine since the write-lock is held only for the brief "flip" of replacing the shared_ptr after the load is complete, not for the whole load operation, right? Though I suppose you'd also be waiting for in-flight reads to complete in order for the write lock to be taken - consider making the read-locks also very short; take a copy of the shared pointer before reading, drop the lock once you have it, and the duration of the read is no longer blocking.

Reply:
... for readability and simplicity you could make all the mutex-guarded operations their own functions, e.g.
```
MaxmindDbSharedPtr getCityDb() ABSL_LOCKS_EXCLUDED(mmdb_mutex_) {
  absl::ReaderMutexLock lock(&mmdb_mutex_);
  return city_db_;
}
void updateCityDb(MaxmindDbSharedPtr city_db) ABSL_LOCKS_EXCLUDED(mmdb_mutex_) {
  absl::MutexLock lock(&mmdb_mutex_);
  city_db_ = city_db;
}
```
That way all the mutex-related operations are constrained to a very small area with very clear semantics (and it's super obvious that it can't deadlock).